u-boot/spi-installer.nix

```
@@ -0,0 +1,107 @@
+{ pkgs
+, e2fsprogs
+, runCommandNoCC
+, uboot
+, utillinux
+, ubootTools
+}:
+
+let
+  board = "pinebook-pro-rk3399";
+
+  flashscript = pkgs.writeText "${board}-flash.cmd" ''
+    echo
+    echo
+    echo ${board} firmware installer
+    echo
+    echo
+
+    if load ''${devtype} ''${devnum}:''${bootpart} ''${kernel_addr_r} ''${board_name}.spiflash.bin; then
+      sf probe
+      sf erase 0 +$filesize
+      sf write $kernel_addr_r 0 $filesize
+      echo "Flashing seems to have been successful! Resetting in 5 seconds"
+      sleep 5
+      reset
+    fi
+
+  '';
+
+  bootcmd = pkgs.writeText "${board}-boot.cmd" ''
+    setenv bootmenu_0 'Flash firmware to SPI=setenv script flash.scr; run boot_a_script'
+    setenv bootmenu_1 'Completely erase SPI=sf probe; echo "Currently erasing..."; sf erase 0 +1000000; echo "Done!"; sleep 5; bootmenu -1'
+    setenv bootmenu_2 'Reboot=reset'
+    setenv bootmenu_3
+    bootmenu -1
+  '';
+
+  mkScript = file: runCommandNoCC "${board}-boot.scr" {
+    nativeBuildInputs = [
+      ubootTools
+    ];
+  } ''
+    mkimage -C none -A arm64 -T script -d ${file} $out
+  '';
+
+  rootfsImage = runCommandNoCC "u-boot-installer-fs.img" {
+    inherit board;
+    size = "8"; # in MiB
+    nativeBuildInputs = [
+      e2fsprogs.bin
+      utillinux
+    ];
+    volumeLabel = "FIRMWARE_INSTALL";
+    uuid = "666efd84-5c25-48ec-af06-e9dadbaa830f";
+  } ''
+    img="$out"
+    (PS4=" $ "; set -x
+
+    fallocate -l "$size"M $img
```

I have this error when building:
```
building '/nix/store/dlijpkmx1ymyq0b6sia5snf5dqcbdp70-u-boot-installer-fs.img.drv'...
  $ fallocate -l 8M /nix/store/d5xyf9f8wnj1jl8jmyir47v7dyak1qb4-u-boot-installer-fs.img
fallocate: fallocate failed: Operation not supported
builder for '/nix/store/dlijpkmx1ymyq0b6sia5snf5dqcbdp70-u-boot-installer-fs.img.drv' failed with exit code 1
```
I'm running on ZFS, but I'm a little surprised since I didn't encounter this last time. Maybe I built it on a different machine or something.
```suggestion
    truncate -s "$size"M $img
```
`fallocate -x -l` works as well

---

Thanks!

That's annoying :/